File: anton/SummerOfCode/src/summerofcode/Day1.java
/*
 * Day 1 parts 1 and 2
 */
package summerofcode;

import java.io.FileReader;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

/**
 *
 * @author Anton Dil
 */
public class Day1
{
   private final static Map<String,Double> scores = new HashMap<>();
   
   static
   {
      scores.put("Almaty", 2.0);
      scores.put("Brorfelde", 0.9);
      scores.put("Estacada", 0.4);
      scores.put("Jayuya", 0.6);
      scores.put("Karlukovo", 2.2);
      scores.put("Morgantown", 2.9);
      scores.put("Nordkapp", 1.5);
      scores.put("Nullarbor", 2.2);
      scores.put("Puente-Laguna-Garzonkuala-Penyu", 0.4);
      scores.put("Uzupis", 0.9);
   }
   
   public static void main(String[] args) throws Exception
   { 
      Scanner sc = new Scanner(new FileReader("01-holidays.txt"));  
      
      String id, loc, bestId = "";
      double value, score;
      double bestValue = 0;      
      int price, days;      
      int afford = 0;
      
      while(sc.hasNext())
      {
         id = sc.next();         
         price = sc.nextInt();         
         loc = sc.next();        
         days = sc.nextInt(); 
         
         if(scores.get(loc) != null)
         {
            score = scores.get(loc);
         }
         else
         {
            score = 1.0;
         }
      
         //only consider holidays we can afford        
         if(price <= 1200)
         {                     
            value = score * days/price;   
         
            System.out.printf("%s %d %s %d value %.5f %n", id, price, loc, days, value);
            afford++;
            
            if(value > bestValue)
            {
               bestValue = value;
               bestId = id;
            }
         }
      }      

      System.out.println("Affordable " + afford);
      System.out.printf("Best value is id %s with value %f%n", bestId, bestValue);
   }
}

File: 01-ticket-prices/day1.rb

$destination_values = {'Almaty' => 2.0, 'Brorfelde' => 0.9, 'Estacada' => 0.4, 
  'Jayuya' => 0.6, 'Karlukovo' => 2.2, 'Morgantown' => 2.9,'Nordkapp' => 1.5, 
  'Nullarbor' => 2.2, 'Puente-Laguna-Garzonkuala-Penyu' => 0.4, 'Uzupis' => 0.9}
$destination_values.default = 1.0

holidays = File.readlines('01-holidays.txt').map {|l| l.chomp.split}

def affordable_holidays(hs)
  hs.select {|h| h[1].to_i <= 1200}
end

def value(holiday)
  $destination_values[holiday[2]] * holiday[3].to_f / holiday[1].to_i 
end

part1 = affordable_holidays(holidays).length
part2 = affordable_holidays(holidays).max_by {|h| value h}.first

puts "Part 1: #{part1}\nPart 2: #{part2}"
